File: Agp2p.Web.UI/Page/loan.cs
using System;
using System.ComponentModel;
using System.Web.Services;
using System.Linq;
using System.Net;
using System.Web;
using Agp2p.Common;
using Agp2p.Linq2SQL;
using Newtonsoft.Json;

namespace Agp2p.Web.UI.Page
{
    /// <summary>
    /// 我要借款
    /// </summary>
    public partial class loan : Web.UI.BasePage
    {
        protected LoanApplyStep step = LoanApplyStep.Unlogin;
        protected int user_id = 0;
        protected string user_name = "";
        protected int loaner_id = 0;
        protected int pending_project_id = 0;
        protected int quota_use = 0;

        protected li_loaners loaner;
        protected li_projects project;

        protected string loaner_name = "";
        protected string loaner_mobile = "";
        protected string loaner_age = "";
        protected string loaner_native_place = "";
        protected string loaner_job = "";
        protected string loaner_working_at = "";
        protected string loaner_working_company = "";
        protected string loaner_educational_background = "";
        protected string loaner_marital_status = "";
        protected string loaner_income = "";

        protected int project_category_id = 0;
        protected int project_amount = 0;
        protected string project_loan_usage = "";
        protected string project_loaner_content = "";
        protected string project_source_of_repayment = "";

        /// <summary>
        /// 重写父类的虚方法,此方法将在Init事件前执行
        /// </summary>
        protected override void ShowPage()
        {
            Init += Page_Init; 
        }

        public enum LoanApplyStep
        {
            [Description("未登录")]
            Unlogin = 0,
            [Description("未申请借款人")]
            UnApplyAsLoaner = 1,
            [Description("申请借款人审核中")]
            ApplyAsLoanerAuditting = 2,
            [Description("未申请借款项目")]
            UnApplyProject = 3,
            [Description("申请借款项目审核中")]
            ProjectAuditting = 4,
            [Description("完成借款申请")]
            ProjectApplyCompleted = 5,
        }

        void Page_Init(object sender, EventArgs e)
        {
            var user = GetUserInfoByLinq();
            if (user == null)
                step = LoanApplyStep.Unlogin;
            else
            {
                user_id = user.id;
                user_name = user.user_name;
                //查看是否已为借款人
                Agp2pDataContext context = new Agp2pDataContext();
                loaner = context.li_loaners.SingleOrDefault(l => l.user_id == user.id);
                if (loaner == null)
                {
                    step = LoanApplyStep.UnApplyAsLoaner;
                    loaner_name = user.real_name;
                    loaner_mobile = user.mobile;
                }
                else
                {
                    //借款人信息
                    loaner_id = loaner.id;
                    loaner_age = loaner.age.ToString();
                    loaner_educational_background = loaner.educational_background;
                    loaner_income = loaner.income;
                    loaner_job = loaner.job;
                    loaner_marital_status = loaner.marital_status.ToString();
                    loaner_mobile = loaner.dt_users.mobile;
                    loaner_name = loaner.dt_users.real_name;
                    loaner_native_place = loaner.native_place;
                    loaner_working_at = loaner.working_at;
                    loaner_working_company = loaner.working_company;

                    if (loaner.status != (int) Agp2pEnums.LoanerStatusEnum.Normal)
                    {
                        step = LoanApplyStep.ApplyAsLoanerAuditting;
                        return;
                    }

                    //查看是否有在审批中的借款申请
                    project = context.li_projects.FirstOrDefault(p => p.li_risks.li_loaners.dt_users.id == user.id &&
                                                                      p.status <= (int) Agp2pEnums.ProjectStatusEnum.FinancingApplicationChecking);
                    if (project != null)
                    {
                        step = LoanApplyStep.ProjectAuditting; //显示正在审批步骤
                        project_category_id = project.category_id;
                        project_amount = (int)project.financing_amount;
                        project_loan_usage = project.li_risks.loan_usage;
                        project_source_of_repayment = project.li_risks.source_of_repayment;
                        project_loaner_content = project.li_risks.loaner_content;
                        quota_use = loaner.quota - (int)project.financing_amount;
                    }
                    else
                    {
                        //查询已审核通过的借款
                        var projectAll = context.li_projects.Where(p =>
                                    p.li_risks.li_loaners.dt_users.id == user.id &&
                                    (int) Agp2pEnums.ProjectStatusEnum.FinancingApplicationSuccess <= p.status &&
                                    p.status != (int) Agp2pEnums.ProjectStatusEnum.FinancingFail);
                        //查询可用额度
                        if (projectAll.Any())
                        {
                            step = LoanApplyStep.ProjectApplyCompleted;
                            quota_use = loaner.quota - (int)projectAll.Sum(p => p.financing_amount);
                        }
                        else
                        {
                            step = LoanApplyStep.UnApplyProject;
                            quota_use = loaner.quota;
                        }
                    }
                }
            }
        }

        /// <summary>
        /// 申请成为借款人
        /// </summary>
        /// <param name="age"></param>
        /// <param name="native_place"></param>
        /// <param name="job"></param>
        /// <param name="working_at"></param>
        /// <param name="working_company"></param>
        /// <param name="educational_background"></param>
        /// <param name="marital_status"></param>
        /// <param name="income"></param>
        /// <returns></returns>
        [WebMethod]
        public static string ApplyLoaner(short age, string nativePlace, string job, string workingCompany, string educationalBackground, int maritalStatus, string income)
        {
            var context = new Agp2pDataContext();
            var userInfo = GetUserInfoByLinq(context);
            HttpContext.Current.Response.TrySkipIisCustomErrors = true;
            if (userInfo == null)
            {
                HttpContext.Current.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                return "请先登录";
            }

            try
            {
                var loaner = context.li_loaners.FirstOrDefault(l => l.dt_users.id == userInfo.id);
                if (loaner != null)
                {
                    loaner.age = age;
                    loaner.native_place = nativePlace;
                    loaner.job = job;
                    loaner.working_at = "";
                    loaner.working_company = workingCompany;
                    loaner.educational_background = educationalBackground;
                    loaner.marital_status = (byte)maritalStatus;
                    loaner.income = income;
                    loaner.status = (int)Agp2pEnums.LoanerStatusEnum.Pending;
                    loaner.last_update_time = DateTime.Now;
                }
                else
                {
                    loaner = new li_loaners()
                    {
                        user_id = userInfo.id,
                        age = age,
                        native_place = nativePlace,
                        job = job,
                        working_at = "",
                        working_company = workingCompany,
                        educational_background = educationalBackground,
                        marital_status = (byte)maritalStatus,
                        income = income,
                        status = (int)Agp2pEnums.LoanerStatusEnum.Pending,
                        last_update_time = DateTime.Now
                    };
                    context.li_loaners.InsertOnSubmit(loaner);
                }
                context.SubmitChanges();
            }
            catch (Exception ex)
            {
                throw ex;
            }

            return "ok";
        }

        /// <summary>
        /// 申请借款
        /// </summary>
        /// <param name="loaner_id"></param>
        /// <param name="user_name"></param>
        /// <param name="loaner_content"></param>
        /// <param name="loan_usage"></param>
        /// <param name="source_of_repayment"></param>
        /// <param name="category_id"></param>
        /// <param name="amount"></param>
        /// <returns></returns>
        [WebMethod]
        public static string ApplyLoan(string loanerContent, string loanUsage, string sourceOfRepayment, decimal amount)
        {
            var context = new Agp2pDataContext();
            var userInfo = GetUserInfoByLinq(context);
            HttpContext.Current.Response.TrySkipIisCustomErrors = true;
            if (userInfo == null)
            {
                HttpContext.Current.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                return "请先登录";
            }

            var project = context.li_projects.FirstOrDefault(p => p.li_risks.li_loaners.dt_users.id == userInfo.id &&
                                                              p.status < (int)Agp2pEnums.ProjectStatusEnum.FinancingApplicationChecking);
            if (project != null)
            {
                project.li_risks.loaner_content = loanerContent;
                project.li_risks.loan_usage = loanUsage;
                project.li_risks.source_of_repayment = sourceOfRepayment;
                project.financing_amount = amount;
                project.status = (int) Agp2pEnums.ProjectStatusEnum.FinancingApplicationUncommitted;

                context.SubmitChanges();
                return "ok";
            }

            try
            {
                var risk = new li_risks()
                {
                    loaner = userInfo.li_loaners.Single().id,
                    loaner_content = loanerContent,
                    loan_usage = loanUsage,
                    source_of_repayment = sourceOfRepayment,
                    last_update_time = DateTime.Now
                };
                project = new li_projects()
                {
                    li_risks = risk,
                    category_id = 61,
                    financing_amount = amount,
                    user_name = userInfo.user_name,
                    add_time = DateTime.Now,
                    status = (int)Agp2pEnums.ProjectStatusEnum.FinancingApplicationUncommitted,
                    repayment_type = (int)Agp2pEnums.ProjectRepaymentTypeEnum.DengEr,
                    no = "",
                    type = (int)Agp2pEnums.LoanTypeEnum.Personal
                };
                //项目编号
                var latestProject = context.li_projects.Where(p => p.category_id == project.category_id).OrderByDescending(p => p.id).FirstOrDefault();
                int prjectCount = latestProject == null ? 0 : Utils.StrToInt(latestProject.title.Substring(latestProject.title.Length - 5), 0);
                project.title += new BLL.article_category().GetModel(project.category_id).call_index.ToUpper() + (prjectCount + 1).ToString("00000");
                context.li_risks.InsertOnSubmit(risk);
                context.li_projects.InsertOnSubmit(project);
                context.SubmitChanges();
            }
            catch (Exception ex)
            {
                throw ex;
            }

            return "ok";
        }
    }
}

File: Agp2p.Web/templates/AnGuang/js/loan.js
import React from "react";
import ReactDom from "react-dom";

import "bootstrap-webpack!./bootstrap.config.js";
import "../less/head.less";
import "../less/loan.less";
import "../less/footerSmall.less";
import alert from "../components/tips_alert.js";

import header from "./header.js";

window['$'] = $;

const LoanApplyStep = {
    Unlogin : 0, // 未登录
    UnApplyAsLoaner : 1, // 未申请借款人
    ApplyAsLoanerAuditting : 2, // 申请借款人审核中
    UnApplyProject : 3, // 未申请借款项目
    ProjectAuditting : 4, // 申请借款项目审核中
    ProjectApplyCompleted : 5, // 完成借款申请
};

const LoanerStatusEnum = {
    IsNotALoaner : -1, // 不是借款人
    Normal : 1, // 正常
    Pending : 10, // 待审核
    PendingFail : 11, // 审核不通过
    Disable : 20 // 禁用
};

// 这个枚举跟服务器上面的不同
const ProjectStatusEnum = {
    FinancingApplicationNotExist : -1, // 未申请
    FinancingApplicationUncommitted : 1, // 提交了申请
    FinancingApplicationChecking : 2, // 审核中
    FinancingApplicationCancel : 0, // 审核失败
};

class ApplyingStatusPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }
    render () {
        var step1HighLightStyle = this.props.step >= 1
            ? {background: 'url("/templates/AnGuang/imgs/loan/personal-info.png") no-repeat'} : null;
        var step2HighLightStyle = this.props.step >= 2
            ? {background: 'url("/templates/AnGuang/imgs/loan/loan.png") no-repeat'} : null;
        var step3HighLightStyle = this.props.step >= 5
            ? {background: 'url("/templates/AnGuang/imgs/loan/loan-finish.png") no-repeat'} : null;

        return (
            <div className="nav-wrap">
                <p className="title">借款流程</p>
                <ul className="list-unstyled list-inline application-ul">
                    <li className="step1">登录</li>
                    <li className="next-icon"></li>
                    <li className="step2" style={step1HighLightStyle} >申请成为借款人</li>
                    <li className="next-icon"></li>
                    <li className="step3" style={step2HighLightStyle} >申请借款</li>
                    <li className="next-icon"></li>
                    <li className="step4" style={step3HighLightStyle} >完成借款</li>
                </ul>
            </div>
        );
    }
}

class LoginPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }
    doLogin() {
        $.ajax({
            type: "post",
            url: "/tools/submit_ajax.ashx?action=user_login",
            dataType: "json",
            data: {
                txtUserName: this.refs.userName.value,
                txtPassword: this.refs.passwd.value,
                chkRemember: true
            },
            success: data => {
                if(data.status == 1){
                    location.reload();
                } else {
                    alert(data.msg);
                }
            },
            error: function(xhr, status, err){
                alert("操作超时，请重试。");
            }
        });
    }
    render () {
        return (
            <div className="login-form-wrap">
                <div className="form-title">
                    <span className="title-style">登录</span>
                    <span className="pull-right register-btn">如无账号 <a href='/register.html'>快速注册</a></span>
                </div>
                <form className="form-horizontal login-form">
                    <div className="form-group">
                        <label htmlFor="user-name" className="control-label">用户名：</label>
                        <input type="text" className="form-control" ref="userName" id="user-name" />
                    </div>
                    <div className="form-group">
                        <label htmlFor="user-pwd" className="control-label">密码：</label>
                        <input type="text" style={{display: 'none'}} />
                        <input type="password" className="form-control" ref="passwd" id="user-pwd" autoComplete="off" />
                    </div>
                    <div className="form-group">
                        <button type="button" className="btn btn-default" id="loginBtn" onClick={ev => this.doLogin()}>登 录</button>
                    </div>
                </form>
            </div>
        );
    }
}

class LoanerApplyingPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = props.value;
    }
    componentDidMount() {
        if(!this.state.loanerName){
            alert("请先前往会员中心进行实名认证", function(){
                location.href = '/user/center/index.html#/safe';
            });
        }
    }
    
    doLoanerApplySubmit() {
        $.ajax({
            type: "post",
            url: "/aspx/main/loan.aspx/ApplyLoaner",
            dataType: "json",
            contentType: "application/json",
            data: JSON.stringify({
                age:0,
                nativePlace: this.state.loanerNativePlace,
                job: this.state.loanerJob,
                workingCompany: this.state.loanerWorkingCompany,
                educationalBackground: this.state.loanerEducationalBackground,
                maritalStatus: this.state.loanerMaritalStatus,
                income: this.state.loanerIncome
            }),
            success: data => {
                if (data.d === "ok") location.reload();
            },
            error: function(xhr, status, err){                
                alert(xhr.responseJSON.Message);
            }
        });
    }
    render () {
        var { loanerName,loanerMobile,loanerAge,loanerEducationalBackground,loanerIncome, loanerStatus,
            loanerJob,loanerMaritalStatus,loanerNativePlace,loanerWorkingAt,loanerWorkingCompany } = this.state;
        var isEditable = loanerStatus == LoanerStatusEnum.IsNotALoaner || loanerStatus == LoanerStatusEnum.PendingFail;
        return (
            <div className="personal-info-form-wrap">
                <div className="form-title">
                    <span className="title-style">借款人资料</span>
                </div>
                {loanerStatus == LoanerStatusEnum.Pending ?
                <div id="loanerApplyChecking" className="status">
                    <span className="status-icon checking"></span>
                    <span className="tips-title">申请借款人审核中</span>
                    <span className="tips">您提交的申请正在审核中，请耐心等待！</span>
                </div> : null }
                {loanerStatus == LoanerStatusEnum.PendingFail ?
                <div id="loanerApplyFailed" className="status">
                    <span className="status-icon failed"></span>
                    <span className="tips-title">申请借款人失败</span>
                    <span className="tips">您提交的申请资料有误，请重新核对资料后再提交！</span>
                </div>:null}
                {loanerStatus == LoanerStatusEnum.Disable ?
                <div id="loanerApplyForbid" className="status">
                    <span className="status-icon forbid"></span>
                    <span className="tips-title">禁止申请借款人</span>
                    <span className="tips">抱歉！您已被禁止申请成为借款人。</span>
                </div>:null}
                <form className="form-horizontal personal-info-form">
                    <div className="form-group">
                        <label htmlFor="name" className="control-label">姓名：</label>
                        <input type="text" id="name" value={loanerName} className="form-control" readOnly />
                    </div>
                    <div className="form-group">
                        <label htmlFor="phone" className="control-label">手机号码：</label>
                        <input type="text" id="phone" value={loanerMobile} className="form-control" readOnly />
                    </div>
                    <div className="form-group">
                        <label htmlFor="birthplace" className="control-label">籍贯：</label>
                        <input type="text" id="birthplace" className="form-control" readOnly={!isEditable}
                            value={loanerNativePlace} onChange={ev => this.setState({loanerNativePlace: ev.target.value})} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="job" className="control-label">职业：</label>
                        <input type="text" id="job" className="form-control" value={loanerJob} readOnly={!isEditable}
                            onChange={ev => this.setState({loanerJob: ev.target.value})} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="employer" className="control-label">工作单位：</label>
                        <input type="text" id="employer" className="form-control" value={loanerWorkingCompany} readOnly={!isEditable}
                            onChange={ev => this.setState({loanerWorkingCompany: ev.target.value})} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="education" className="control-label">学历：</label>
                        <select id="education" className="form-control" value={loanerEducationalBackground} disabled={!isEditable}
                            onChange={ev => this.setState({loanerEducationalBackground: ev.target.value})} >
                            <option value="" style={{display: 'none'}}></option>
                            <option value="小学及以下">小学及以下</option>
                            <option value="初中">初中</option>
                            <option value="高中">高中</option>
                            <option value="中专">中专</option>
                            <option value="大专">大专</option>
                            <option value="本科">本科</option>
                            <option value="研究生">研究生</option>
                            <option value="博士及以上">博士及以上</option>
                        </select>
                    </div>
                    <div className="form-group">
                        <label htmlFor="marital-status" className="control-label">婚姻状况：</label>
                        <select id="marital-status" className="form-control" value={loanerMaritalStatus} disabled={!isEditable}
                            onChange={ev => this.setState({loanerMaritalStatus: ev.target.value})} >
                            <option style={{display: 'none'}}></option>
                            <option value="1">未婚</option>
                            <option value="2">已婚</option>
                            <option value="3">离异</option>
                            <option value="4">丧偶</option>
                        </select>
                    </div>
                    <div className="form-group">
                        <label htmlFor="income" className="control-label">收入：</label>
                        <input type="text" id="income" className="form-control" value={loanerIncome} readOnly={!isEditable}
                            onChange={ev => this.setState({loanerIncome: ev.target.value})} />
                    </div>
                    <div className="form-group">
                        <button type="button" className="btn btn-default" id="loanerApplyBtn" style={isEditable?null:{display:'none'}} onClick={ev => this.doLoanerApplySubmit()}>提 交</button>
                    </div>
                </form>
            </div>
        );
    }
}

class ProjectApplyingPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = props.value;
        this.state.quotaUse = parseFloat(props.quotaUse);
    }
    doProjectApply() {
        $.ajax({
            type: "post",
            url: "/aspx/main/loan.aspx/ApplyLoan",
            dataType: "json",
            contentType: "application/json",
            data: JSON.stringify({
                loanerContent: this.state.projectLoanerContent,
                loanUsage: this.state.projectLoanUsage,
                sourceOfRepayment: this.state.projectSourceOfRepayment,
                amount: this.state.projectAmount,
            }),
            success: function(data){
                if (data.d === "ok") location.reload();
            },
            error: function(xhr, status, err){                
                alert(xhr.responseJSON.Message);
            }
        });
    }
    doQuotaCheck() {
        if (parseFloat(this.state.projectAmount) > this.state.quotaUse) {
            alert("借款额度不能大于可用额度！");
        }
    }
    render() {
        var { projectCategoryId,projectAmount,projectLoanUsage,projectSourceOfRepayment,projectLoanerContent,quotaUse, projectStatus } = this.state;
        var isEditable = projectStatus == ProjectStatusEnum.FinancingApplicationNotExist || projectStatus == ProjectStatusEnum.FinancingApplicationCancel;
        return (
            <div className="loan-detail-form-wrap">
                <div className="form-title">
                    <span className="title-style">申请借款</span>
                </div>
                {projectStatus != ProjectStatusEnum.FinancingApplicationChecking && projectStatus != ProjectStatusEnum.FinancingApplicationUncommitted
                ? null
                : <div id="loanApplyChecking" className="status">
                    <span className="status-icon checking"></span>
                    <span className="tips-title">申请借款审核中</span>
                    <span className="tips">您提交的申请正在审核中，请耐心等待！</span>
                </div>}
                {projectStatus != ProjectStatusEnum.FinancingApplicationCancel
                ? null
                : <div id="loanApplyFailed" className="status">
                    <span className="status-icon failed"></span>
                    <span className="tips-title">申请借款失败</span>
                    <span className="tips">您提交的申请资料有误，请重新核对资料后再提交！</span>
                </div>}
                <form className="form-horizontal loan-detail-form">
                    <div className="form-group">
                        <label htmlFor="loan-amount" className="control-label">借款金额：</label>
                        <input type="text" id="loan-amount" className="form-control" value={projectAmount}
                            onChange={ev => this.setState({projectAmount: ev.target.value})}
                            readOnly={!isEditable}
                            onBlur={ev => this.doQuotaCheck()}/>
                        <span id="largest-amount">{`（可用额度 ${quotaUse}）`}</span>
                    </div>
                    <div className="form-group">
                        <label htmlFor="loan-description" className="control-label">借款描述：</label>
                        <textarea id="loan-description" className="form-control" value={projectLoanerContent}
                            readOnly={!isEditable}
                            onChange={ev => this.setState({projectLoanerContent: ev.target.value})}></textarea>
                    </div>
                    <div className="form-group">
                        <label htmlFor="loan-usage" className="control-label">借款用途：</label>
                        <textarea id="loan-usage" className="form-control" value={projectLoanUsage}
                            readOnly={!isEditable}
                            onChange={ev => this.setState({projectLoanUsage: ev.target.value})}></textarea>
                    </div>
                    <div className="form-group">
                        <label htmlFor="repayment-source" className="control-label">还款来源：</label>
                        <textarea id="repayment-source" className="form-control" value={projectSourceOfRepayment}
                            readOnly={!isEditable}
                            onChange={ev => this.setState({projectSourceOfRepayment: ev.target.value})}></textarea>
                    </div>
                    <div className="form-group">
                        <button type="button" className="btn btn-default" id="loanApplyBtn"
                            style={isEditable?null:{display: 'none'}}
                            onClick={ev => this.doProjectApply()}>提 交</button>
                    </div>
                </form>
            </div>
        );
    }
}

class CompletePanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }
    render () {
        return (
            <div className="Completion-loan-form-wrap">
                <div className="form-title"><span className="title-style">完成借款</span></div>
                <form className="form-horizontal Completion-loan-form">
                    <div className="form-group">
                        <label className="control-label"><img src={require('../imgs/loan/001.png')} /></label><br /><br />
                        <label className="control-label">恭喜您！您申请的借款已审核通过！</label><br /><br /><br />
                        <label className="control-label"><input id="loanApplyAgainBtn" type="button" value="申请借款" onClick={ev => this.props.onApplyLoanClick()} /></label>
                    </div>
                </form>
            </div>
        );
    }
}

class LoanApplying extends React.Component {
    constructor(props) {
        super(props);
        var { step,userId,userName,loanerId,pendingProjectId,quotaUse,status } = $("#main").data();
        this.state = {step: parseInt(step)};
    }
    render () {
        var step = this.state.step;
        return (
            <div className="content-wrap application-form">
                <ApplyingStatusPanel step={this.state.step} />
                <div className="form-wrapper">
                {step == 0 ? <LoginPanel /> : null}
                {step == 1 || step == 2 ? <LoanerApplyingPanel value={$("#loaner").data()} /> : null}
                {step == 3 || step == 4 ? <ProjectApplyingPanel value={$("#project").data()} quotaUse={$("#main").data().quotaUse} /> : null}
                {step == 5 ? <CompletePanel onApplyLoanClick={() => this.setState({step: 3})} /> : null}
                </div>
            </div>
        );
    }
}

$(function () {
    header.setHeaderHighlight(2);

    //data-toggle 初始化
    $('[data-toggle="popover"]').popover();

    ReactDom.render(<LoanApplying />, document.querySelector('.react-root'));
});

